Remove FutureMessage from RPC request callback logic

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50236 Apply clang-format to rpc cpp files
* #50029 Completely remove FutureMessage type
* #50028 Completely Remove FutureMessage from RPC agents
* #50027 Completely Remove FutureMessage from RPC cpp tests
* **#50026 Remove FutureMessage from RPC request callback logic**
* #50025 Completely remove FutureMessage from FaultyProcessGroupAgent
* #50024 Remove FutureMessage from sender TensorPipeAgent
* #50023 Remove FutureMessage from sender ProcessGroupAgent
* #50020 Completely remove FutureMessage from distributed autograd
* #50005 Completely remove FutureMessage from RPC TorchScript implementations
* #50004 Completely remove FutureMessage from RRef Implementations
* #49995 Replace FutureMessage with ivalue::Future in RpcAgent retry logic
* #49960 Replace FutureMessage with ivalue::Future in RRefContext
* #49927 Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*
* #49906 Let RpcAgent::send() return JitFuture

Differential Revision: [D25753588](https://our.internmc.facebook.com/intern/diff/D25753588)

mrshenli committed:
Remove FutureMessage from RPC request callback logic

[ghstack-poisoned]

mrshenli committed:
Update on "Remove FutureMessage from RPC request callback logic"


Differential Revision: [D25753588](https://our.internmc.facebook.com/intern/diff/D25753588)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/request_callback_no_python.cpp:
```diff
@@ -398,22 +402,24 @@ void RequestCallbackNoPython::processForwardAutogradReq(
       auto msg = getMessageWithAutograd(
           fromWorkerId,
-          std::move(*wrappedRpcResponseFuture).moveValue(),
+          //std::move(*wrappedRpcResponseFuture).moveValue(),
```
Is this comment a leftover?

lw commented on torch/csrc/distributed/rpc/request_callback_no_python.cpp:
```diff
@@ -525,17 +533,18 @@ void RequestCallbackNoPython::processRunWithProfilingReq(
             auto rpcWithProfilingResp = std::make_unique<RpcWithProfilingResp>(
                 MessageType::RUN_WITH_PROFILING_RESP,
-                std::move(*wrappedRpcResponseFuture).moveValue(),
+                //std::move(*wrappedRpcResponseFuture).value(),
```
Same here

lw commented on torch/csrc/distributed/rpc/request_callback_no_python.cpp:
```diff
@@ -629,7 +639,7 @@ void RequestCallbackNoPython::processRpc(
 }
 
-Message RequestCallbackNoPython::handleError(
+IValue RequestCallbackNoPython::handleError(
```
What's the reason for changing this return type? It doesn't seem to be strictly necessary...

mrshenli replied:
`handleError` was used in multiple places like this:

```
retFuture->markCompleted(handleError(e, request.type(), request.id()));
```

We can do the conversion in on the callsite, but doing it within `handleError` can help dedup some conversion code.

mrshenli committed:
Update on "Remove FutureMessage from RPC request callback logic"


Differential Revision: [D25753588](https://our.internmc.facebook.com/intern/diff/D25753588)

[ghstack-poisoned]

mrshenli committed:
Update on "Remove FutureMessage from RPC request callback logic"


Differential Revision: [D25753588](https://our.internmc.facebook.com/intern/diff/D25753588)

[ghstack-poisoned]
